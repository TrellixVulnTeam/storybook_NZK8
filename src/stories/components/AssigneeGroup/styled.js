// @flow

import styled from 'styled-components';
import { colors } from 'utils/StyleGuide';

export const AssigneeWrapper = styled.div`
  display: flex;
  align-items: center;

  ul {
    display: inline-flex;
    padding: 0;
    margin: 0;
  }

  /* NumberBadge */
  span {
    position: relative;
    left: -24px;
    top: 10px;
    z-index: 10;

    @media (max-width: 1445px) and (min-width: 1200px) {
      display: none;
    }

    @media (max-width: 1185px) {
      display: none;
    }
  }
`;

export const Assignee = styled.li`
  width: ${p => (p.size === 'sm' ? '24px' : '36px')};
  height: ${p => (p.size === 'sm' ? '24px' : '36px')};
  line-height: ${p => (p.size === 'sm' ? '22px' : '36px')};
  text-align: center;
  font-size: ${p => p.size === 'sm' && '10px'};
  display: inline-block;
  border: 1px solid ${colors.white};
  background-image: url(${props => props.avatar});
  background-position: center;
  background-size: cover;
  background-repeat: no-repeat;
  background-origin: border-box;
  background-color: ${props => props.assigneeColor};
  color: ${colors.white} !important;
  border-radius: 50%;

  &:first-of-type {
    z-index: ${props => props.offset && props.offset};
  }

  &:not(:first-of-type) {
    transform: translateX(${props => props.offset * props.index * -1}px);
    z-index: ${props => props.offset && (props.index - props.offset) * -1};

    @media (max-width: 1350px) and (min-width: 1200px) {
      display: none;
    }

    @media (max-width: 1090px) {
      display: none;
    }
  }

  svg {
    position: relative;
    top: -1px;
    left: -1px;
    width: 36px;
    height: 36px;
  }
`;
