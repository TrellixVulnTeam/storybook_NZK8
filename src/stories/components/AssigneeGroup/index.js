// @flow
import React from 'react';
import NumberBadge from 'components/NumberBadge';
import Tooltip from 'components/Tooltip';
import { generateUserInitials, generateUserName, generateUserColor } from 'utils';
import type { User, Team } from 'types';
import { AssigneeWrapper, Assignee } from './styled';
import TeamIcon from './img/team.svg';

type Props = {
  id: number | string,
  assignedTo?: [User],
  supportUsers?: User[],
  team?: Team,
  threshold: number,
  offset?: number,
  size?: 'sm',
  tooltip?: 'true' | 'false',
};

type AssigneeType = $Shape<User & { team?: boolean }>;

export default (props: Props) => {
  const { id, assignedTo, supportUsers, team, threshold, offset, size, tooltip } = props;
  const assignees: AssigneeType[] = [];

  if (assignedTo) {
    assignees.push(assignedTo[0]);
  }

  if (supportUsers) {
    supportUsers.forEach(assignee => {
      assignees.push(assignee);
    });
  }

  if (team && team[0]) {
    assignees.push({
      team: true,
      ...team[0],
    });
  }

  /* eslint-disable no-shadow */
  /**
   * @param {AssigneeType[]} assignees
   * @param {number} threshold
   * @return {React$Node}
   */
  const truncateAssignees = (assignees: AssigneeType[], threshold: number) => {
    const truncatedAssignees = assignees.slice(0, threshold);
    return (
      <AssigneeWrapper>
        <ul>
          {// eslint-disable-next-line array-callback-return
          truncatedAssignees.map((assignee, idx) => {
            if (assignee) {
              const assigneeInitials = generateUserInitials(assignee);
              const assigneeColor = generateUserColor(assignee.id);
              const assigneeContent = assignee.avatar ? '' : assigneeInitials.toUpperCase();
              return (
                <Assignee
                  size={size}
                  offset={offset}
                  index={idx}
                  avatar={assignee.avatar}
                  assigneeColor={assigneeColor}
                  // eslint-disable-next-line react/no-array-index-key
                  key={`${assigneeInitials}-${id}-${idx}`}
                >
                  {assignee.team ? <TeamIcon /> : assigneeContent}
                </Assignee>
              );
            }
          })}
        </ul>
        {assignees.length > threshold && <NumberBadge plus count={assignees.length - threshold} />}
      </AssigneeWrapper>
    );
  };
  /* eslint-enable no-shadow */

  // eslint-disable-next-line array-callback-return
  const tooltipContent = assignees.map((assignee, idx) => {
    const assigneeName = generateUserName(assignee);

    // First assginee and not a team
    if (idx === 0 && assignee && !assignee.hasOwnProperty('team')) {
      return assigneeName;
    }

    // Not first assginee and not a team
    if (idx !== 0 && assignee && !assignee.hasOwnProperty('team')) {
      return `<br />${assigneeName}`;
    }

    // Many assignees with a team
    if (idx !== 0 && assignee && assignee.hasOwnProperty('team')) {
      return `<br /><aside></aside><i>Team</i>${assigneeName}`;
    }

    // Only a team
    if (idx === 0 && assignee && assignee.hasOwnProperty('team')) {
      return `<i>Team</i>${assigneeName}`;
    }
  });

  return (
    <Tooltip
      id={`tooltip-${id}`}
      value={tooltipContent}
      // eslint-disable-next-line react/no-children-prop
      children={!assignees.length ? 'No Assignee(s)' : truncateAssignees(assignees, threshold)}
      effect="solid"
      place="bottom"
      display={tooltip}
    />
  );
};
