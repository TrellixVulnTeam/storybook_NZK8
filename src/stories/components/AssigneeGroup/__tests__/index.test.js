import React from 'react';
import { mount } from 'enzyme';
import AssigneeGroup from 'components/AssigneeGroup/index';

describe('<AssigneeGroup />', () => {
  const mandatoryProps = {
    id: 1,
    threshold: 1,
  };

  it('should render correctly with all params given', () => {
    const props = {
      ...mandatoryProps,
      assignedTo: [{ id: 'id', avatar: 'avatar' }],
      supportUsers: [],
      team: [],
      offset: 1,
      size: 'sm',
      tooltip: 'true',
    };
    const sut = mount(<AssigneeGroup {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly with only mandatory params given', () => {
    const sut = mount(<AssigneeGroup {...mandatoryProps} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly for assignee and team', () => {
    const props = {
      ...mandatoryProps,
      assignedTo: [{ id: 'id' }],
      team: ['SomeTeam'],
    };
    const sut = mount(<AssigneeGroup {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly for a team', () => {
    const props = {
      ...mandatoryProps,
      assignedTo: [{ id: 'id', team: 'SomeTeam' }],
    };
    const sut = mount(<AssigneeGroup {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly for many assignees', () => {
    const props = {
      ...mandatoryProps,
      assignedTo: [{ id: 'id' }],
      supportUsers: [{ id: 'id2' }],
    };
    const sut = mount(<AssigneeGroup {...props} />);

    expect(sut).toMatchSnapshot();
  });
});
