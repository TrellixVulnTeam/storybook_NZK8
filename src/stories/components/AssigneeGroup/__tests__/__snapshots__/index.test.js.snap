// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssigneeGroup /> should render correctly for a team 1`] = `
<_default
  assignedTo={
    Array [
      Object {
        "id": "id",
        "team": "SomeTeam",
      },
    ]
  }
  id={1}
  threshold={1}
>
  <_default
    effect="solid"
    id="tooltip-1"
    place="bottom"
    value={
      Array [
        "<i>Team</i> ",
      ]
    }
  >
    <styled__Wrapper
      data-for="tooltip-1"
      data-tip={
        Array [
          "<i>Team</i> ",
        ]
      }
      effect="solid"
      id="tooltip-1"
      place="bottom"
      value={
        Array [
          "<i>Team</i> ",
        ]
      }
    >
      <div
        className="styled__Wrapper-funpqb-0 cznzQQ"
        data-for="tooltip-1"
        data-tip={
          Array [
            "<i>Team</i> ",
          ]
        }
        id="tooltip-1"
        value={
          Array [
            "<i>Team</i> ",
          ]
        }
      >
        <styled__AssigneeWrapper>
          <div
            className="styled__AssigneeWrapper-sc-1sogfbk-0 liVcjM"
          >
            <ul>
              <styled__Assignee
                assigneeColor="#801e2c"
                index={0}
                key="T-1-0"
              >
                <li
                  className="styled__Assignee-sc-1sogfbk-1 jhRkyk"
                >
                  <test-file-stub />
                </li>
              </styled__Assignee>
            </ul>
          </div>
        </styled__AssigneeWrapper>
        <styled__TooltipWrapper
          display="true"
        >
          <div
            className="styled__TooltipWrapper-funpqb-1 gQtIVd"
            display="true"
          >
            <ReactTooltip
              clickable={false}
              delayShow={300}
              effect="solid"
              getContent={[Function]}
              html={true}
              id="tooltip-1"
              insecure={true}
              multiline={true}
              place="bottom"
              resizeHide={true}
              type="dark"
              wrapper="div"
            >
              <div
                className="__react_component_tooltip place-bottom type-dark"
                data-id="tooltip"
                id="tooltip-1"
              >
                null
              </div>
            </ReactTooltip>
          </div>
        </styled__TooltipWrapper>
      </div>
    </styled__Wrapper>
  </_default>
</_default>
`;

exports[`<AssigneeGroup /> should render correctly for assignee and team 1`] = `
<_default
  assignedTo={
    Array [
      Object {
        "id": "id",
      },
    ]
  }
  id={1}
  team={
    Array [
      "SomeTeam",
    ]
  }
  threshold={1}
>
  <_default
    effect="solid"
    id="tooltip-1"
    place="bottom"
    value={
      Array [
        " ",
        "<br /><aside></aside><i>Team</i> ",
      ]
    }
  >
    <styled__Wrapper
      data-for="tooltip-1"
      data-tip={
        Array [
          " ",
          "<br /><aside></aside><i>Team</i> ",
        ]
      }
      effect="solid"
      id="tooltip-1"
      place="bottom"
      value={
        Array [
          " ",
          "<br /><aside></aside><i>Team</i> ",
        ]
      }
    >
      <div
        className="styled__Wrapper-funpqb-0 cznzQQ"
        data-for="tooltip-1"
        data-tip={
          Array [
            " ",
            "<br /><aside></aside><i>Team</i> ",
          ]
        }
        id="tooltip-1"
        value={
          Array [
            " ",
            "<br /><aside></aside><i>Team</i> ",
          ]
        }
      >
        <styled__AssigneeWrapper>
          <div
            className="styled__AssigneeWrapper-sc-1sogfbk-0 liVcjM"
          >
            <ul>
              <styled__Assignee
                assigneeColor="#801e2c"
                index={0}
                key="-1-0"
              >
                <li
                  className="styled__Assignee-sc-1sogfbk-1 jhRkyk"
                />
              </styled__Assignee>
            </ul>
            <_default
              count={1}
              plus={true}
            >
              <styled__NumberBadgeWrapper>
                <span
                  className="styled__NumberBadgeWrapper-sc-2rr4cw-0 hgFfEW"
                >
                  +
                  1
                </span>
              </styled__NumberBadgeWrapper>
            </_default>
          </div>
        </styled__AssigneeWrapper>
        <styled__TooltipWrapper
          display="true"
        >
          <div
            className="styled__TooltipWrapper-funpqb-1 gQtIVd"
            display="true"
          >
            <ReactTooltip
              clickable={false}
              delayShow={300}
              effect="solid"
              getContent={[Function]}
              html={true}
              id="tooltip-1"
              insecure={true}
              multiline={true}
              place="bottom"
              resizeHide={true}
              type="dark"
              wrapper="div"
            >
              <div
                className="__react_component_tooltip place-bottom type-dark"
                data-id="tooltip"
                id="tooltip-1"
              >
                null
              </div>
            </ReactTooltip>
          </div>
        </styled__TooltipWrapper>
      </div>
    </styled__Wrapper>
  </_default>
</_default>
`;

exports[`<AssigneeGroup /> should render correctly for many assignees 1`] = `
<_default
  assignedTo={
    Array [
      Object {
        "id": "id",
      },
    ]
  }
  id={1}
  supportUsers={
    Array [
      Object {
        "id": "id2",
      },
    ]
  }
  threshold={1}
>
  <_default
    effect="solid"
    id="tooltip-1"
    place="bottom"
    value={
      Array [
        " ",
        "<br /> ",
      ]
    }
  >
    <styled__Wrapper
      data-for="tooltip-1"
      data-tip={
        Array [
          " ",
          "<br /> ",
        ]
      }
      effect="solid"
      id="tooltip-1"
      place="bottom"
      value={
        Array [
          " ",
          "<br /> ",
        ]
      }
    >
      <div
        className="styled__Wrapper-funpqb-0 cznzQQ"
        data-for="tooltip-1"
        data-tip={
          Array [
            " ",
            "<br /> ",
          ]
        }
        id="tooltip-1"
        value={
          Array [
            " ",
            "<br /> ",
          ]
        }
      >
        <styled__AssigneeWrapper>
          <div
            className="styled__AssigneeWrapper-sc-1sogfbk-0 liVcjM"
          >
            <ul>
              <styled__Assignee
                assigneeColor="#801e2c"
                index={0}
                key="-1-0"
              >
                <li
                  className="styled__Assignee-sc-1sogfbk-1 jhRkyk"
                />
              </styled__Assignee>
            </ul>
            <_default
              count={1}
              plus={true}
            >
              <styled__NumberBadgeWrapper>
                <span
                  className="styled__NumberBadgeWrapper-sc-2rr4cw-0 hgFfEW"
                >
                  +
                  1
                </span>
              </styled__NumberBadgeWrapper>
            </_default>
          </div>
        </styled__AssigneeWrapper>
        <styled__TooltipWrapper
          display="true"
        >
          <div
            className="styled__TooltipWrapper-funpqb-1 gQtIVd"
            display="true"
          >
            <ReactTooltip
              clickable={false}
              delayShow={300}
              effect="solid"
              getContent={[Function]}
              html={true}
              id="tooltip-1"
              insecure={true}
              multiline={true}
              place="bottom"
              resizeHide={true}
              type="dark"
              wrapper="div"
            >
              <div
                className="__react_component_tooltip place-bottom type-dark"
                data-id="tooltip"
                id="tooltip-1"
              >
                null
              </div>
            </ReactTooltip>
          </div>
        </styled__TooltipWrapper>
      </div>
    </styled__Wrapper>
  </_default>
</_default>
`;

exports[`<AssigneeGroup /> should render correctly with all params given 1`] = `
<_default
  assignedTo={
    Array [
      Object {
        "avatar": "avatar",
        "id": "id",
      },
    ]
  }
  id={1}
  offset={1}
  size="sm"
  supportUsers={Array []}
  team={Array []}
  threshold={1}
  tooltip="true"
>
  <_default
    display="true"
    effect="solid"
    id="tooltip-1"
    place="bottom"
    value={
      Array [
        " ",
      ]
    }
  >
    <styled__Wrapper
      data-for="tooltip-1"
      data-tip={
        Array [
          " ",
        ]
      }
      display="true"
      effect="solid"
      id="tooltip-1"
      place="bottom"
      value={
        Array [
          " ",
        ]
      }
    >
      <div
        className="styled__Wrapper-funpqb-0 cznzQQ"
        data-for="tooltip-1"
        data-tip={
          Array [
            " ",
          ]
        }
        display="true"
        id="tooltip-1"
        value={
          Array [
            " ",
          ]
        }
      >
        <styled__AssigneeWrapper>
          <div
            className="styled__AssigneeWrapper-sc-1sogfbk-0 liVcjM"
          >
            <ul>
              <styled__Assignee
                assigneeColor="#801e2c"
                avatar="avatar"
                index={0}
                key="-1-0"
                offset={1}
                size="sm"
              >
                <li
                  className="styled__Assignee-sc-1sogfbk-1 cySUzN"
                  offset={1}
                  size="sm"
                />
              </styled__Assignee>
            </ul>
          </div>
        </styled__AssigneeWrapper>
        <styled__TooltipWrapper
          display="true"
        >
          <div
            className="styled__TooltipWrapper-funpqb-1 gQtIVd"
            display="true"
          >
            <ReactTooltip
              clickable={false}
              delayShow={300}
              effect="solid"
              getContent={[Function]}
              html={true}
              id="tooltip-1"
              insecure={true}
              multiline={true}
              place="bottom"
              resizeHide={true}
              type="dark"
              wrapper="div"
            >
              <div
                className="__react_component_tooltip place-bottom type-dark"
                data-id="tooltip"
                id="tooltip-1"
              >
                null
              </div>
            </ReactTooltip>
          </div>
        </styled__TooltipWrapper>
      </div>
    </styled__Wrapper>
  </_default>
</_default>
`;

exports[`<AssigneeGroup /> should render correctly with only mandatory params given 1`] = `
<_default
  id={1}
  threshold={1}
>
  <_default
    effect="solid"
    id="tooltip-1"
    place="bottom"
    value={Array []}
  >
    <styled__Wrapper
      data-for="tooltip-1"
      data-tip={Array []}
      effect="solid"
      id="tooltip-1"
      place="bottom"
      value={Array []}
    >
      <div
        className="styled__Wrapper-funpqb-0 cznzQQ"
        data-for="tooltip-1"
        data-tip={Array []}
        id="tooltip-1"
        value={Array []}
      >
        No Assignee(s)
        <styled__TooltipWrapper
          display="true"
        >
          <div
            className="styled__TooltipWrapper-funpqb-1 gQtIVd"
            display="true"
          >
            <ReactTooltip
              clickable={false}
              delayShow={300}
              effect="solid"
              getContent={[Function]}
              html={true}
              id="tooltip-1"
              insecure={true}
              multiline={true}
              place="bottom"
              resizeHide={true}
              type="dark"
              wrapper="div"
            >
              <div
                className="__react_component_tooltip place-bottom type-dark"
                data-id="tooltip"
                id="tooltip-1"
              >
                null
              </div>
            </ReactTooltip>
          </div>
        </styled__TooltipWrapper>
      </div>
    </styled__Wrapper>
  </_default>
</_default>
`;
